cloud/common.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2024 by Red Hat, Inc. All rights reserved.
+// Use of this source is goverend by the Apache License
+// that can be found in the LICENSE file.
+
+package cloud
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+	"syscall"
+)
+
+// HTTPClient make it easier to swap out the client socket for testing
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewClient initializes the values of the cloud API client configuration
+// used to query the server.
+func NewClient(ctx context.Context, socket HTTPClient, socketPath string) Client {
+	// TODO
+	// - check for valid server path
+	return Client{
+		ctx:        ctx,
+		socket:     socket,
+		socketPath: socketPath,
+		protocol:   "http",
+		host:       "localhost",
+		rawFunc:    func(string, string, int, []byte) {},
+	}
+}
+
+// InitClientUnixSocket configures the client to use a unix domain socket
+// This configures the cloud.Client with the socket path
+// It must be called before using any of the cloud.Client functions.
+func InitClientUnixSocket(ctx context.Context, socketPath string) Client {
+	socket := &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", socketPath)
+			},
+		},
+	}
+	return NewClient(ctx, socket, socketPath)
+}
+
+// Client contains details about the cloud API server connection
+// as well as functions to interact with the server
+type Client struct {
+	ctx        context.Context
+	socket     HTTPClient
+	protocol   string // defaults to http
+	host       string // defaults to localhost
+	socketPath string
+	rawFunc    func(string, string, int, []byte) // Pass the raw json data to a user function
+	Test       bool                              // Used to fake the presense of the socket for testing
+}
+
+// SetRawCallback sets a function that will be called with the server response
+// It is passed the response's method, path, result status, and body bytes
+func (c *Client) SetRawCallback(f func(string, string, int, []byte)) {
+	c.rawFunc = f
+}
+
+// RawURL returns the full url for a route
+func (c Client) RawURL(route string) string {
+	if route[0] == '/' {
+		route = route[1:]
+	}
+	return fmt.Sprintf("%s://%s/%s", c.protocol, c.host, route)
```

Comment:
I prefer to use `net/url`'s `URL.String()` here.

Reply:
That would work too, but I prefer to leave it for now. The protocol/host code needs a refactor in a later commit -- it was never hooked up to the cmdline and is always http and localhost. I think we're always going to be using domain sockets for this so they can go away at some point.